src/vm/contexts.rs

```diff
@@ -63,6 +83,246 @@ pub struct CallStack {
+            .unwrap(); // should always exist, because of checked insert above.
+
+        if principal_map.contains_key(&asset) {
+            principal_map.get_mut(&asset).unwrap().push(transfered); 
+        } else {
+            principal_map.insert(asset, vec![transfered]); 
+        }
+    }
+
+    pub fn add_token_transfer(&mut self, principal: &PrincipalData, asset: AssetIdentifier, amount: i128) -> Result<()> {
+        if amount < 0 {
+            panic!("Should never attempt to log a negative transfer.");
+        }
+
+        let next_amount = self.get_next_amount(principal, &asset, amount)?;
+
+        if !self.token_map.contains_key(principal) {
+            self.token_map.insert(principal.clone(), HashMap::new());
+        }
+
+        let principal_map = self.token_map.get_mut(principal)
+            .unwrap(); // should always exist, because of checked insert above.
+
+        principal_map.insert(asset, next_amount);
+
+        Ok(())
+    }
+
+    // This will add any asset transfer data from other to self,
+    //   aborting _all_ changes in the event of an error, leaving self unchanged
+    pub fn commit_other(&mut self, mut other: AssetMap) -> Result<()> {
+        let mut to_add = Vec::new();
+        for (principal, mut principal_map) in other.token_map.drain() {
+            for (asset, amount) in principal_map.drain() {
+                let next_amount = self.get_next_amount(&principal, &asset, amount)?;
+                to_add.push((principal.clone(), asset, next_amount));
+            }
+        }
+
+        // After this point, this function will not fail.
+        for (principal, mut principal_map) in other.asset_map.drain() {
+            for (asset, mut transfers) in principal_map.drain() {
+                if !self.asset_map.contains_key(&principal) {
+                    self.asset_map.insert(principal.clone(), HashMap::new());
+                }
+
+                let landing_map = self.asset_map.get_mut(&principal)
+                    .unwrap(); // should always exist, because of checked insert above.
+                if landing_map.contains_key(&asset) {
+                    let landing_vec = landing_map.get_mut(&asset).unwrap();
+                    landing_vec.append(&mut transfers);
+                } else {
+                    landing_map.insert(asset, transfers);
+                }
+            }
+        }
+
+
+        for (principal, asset, amount) in to_add.drain(..) {
+            if !self.token_map.contains_key(&principal) {
+                self.token_map.insert(principal.clone(), HashMap::new());
+            }
+
+            let principal_map = self.token_map.get_mut(&principal)
+                .unwrap(); // should always exist, because of checked insert above.
+            principal_map.insert(asset, amount);
+        }
+
+        Ok(())
+    }
+
+    pub fn to_table(mut self) -> HashMap<PrincipalData, HashMap<AssetIdentifier, AssetMapEntry>> {
```

Comment:
Is this only used for testing?  If so, should it have `#[cfg(test)]`?

Reply:
Added in 457c403c